debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java

```diff
@@ -83,43 +87,32 @@ public void dispatchDataChangeEvent(T dataCollectionId, Supplier<ChangeRecordEmi
             throw new IllegalArgumentException("No metadata registered for captured table " + dataCollectionId);
         }
 
-        changeRecordEmitter.get().emitChangeRecords(
+        changeRecordEmitter.emitChangeRecords(
             dataCollectionSchema,
-            new ChangeRecordReceiver(dataCollectionId, changeEventCreator, dataCollectionSchema)
+            receiver
         );
     }
 
-    public void dispatchSchemaChangeEvent(T dataCollectionId, Supplier<SchemaChangeEventEmitter> schemaChangeEventEmitter) throws InterruptedException {
+    public void dispatchSchemaChangeEvent(T dataCollectionId, SchemaChangeEventEmitter schemaChangeEventEmitter) throws InterruptedException {
         if(!filter.isIncluded(dataCollectionId)) {
             LOGGER.trace("Skipping data change event for {}", dataCollectionId);
             return;
         }
 
-        schemaChangeEventEmitter.get().emitSchemaChangeEvent(new SchemaChangeEventReceiver());
+        schemaChangeEventEmitter.emitSchemaChangeEvent(new SchemaChangeEventReceiver());
     }
 
-    private final class ChangeRecordReceiver implements ChangeRecordEmitter.Receiver {
-
-        private final T dataCollectionId;
-        private final ChangeEventCreator changeEventCreator;
-        private final DataCollectionSchema dataCollectionSchema;
-
-        private ChangeRecordReceiver(T dataCollectionId, ChangeEventCreator changeEventCreator,
-                DataCollectionSchema dataCollectionSchema) {
-            this.dataCollectionId = dataCollectionId;
-            this.changeEventCreator = changeEventCreator;
-            this.dataCollectionSchema = dataCollectionSchema;
-        }
+    private final class StreamingChangeRecordReceiver implements ChangeRecordEmitter.Receiver {
 
         @Override
-        public void changeRecord(Operation operation, Object key, Struct value, OffsetContext offsetContext) throws InterruptedException {
+        public void changeRecord(DataCollectionSchema dataCollectionSchema, Operation operation, Object key, Struct value, OffsetContext offsetContext) throws InterruptedException {
             Objects.requireNonNull(key, "key must not be null");
             Objects.requireNonNull(value, "key must not be null");
 
             LOGGER.trace( "Received change record for {} operation on key {}", operation, key);
 
             Schema keySchema = dataCollectionSchema.keySchema();
-            String topicName = topicSelector.topicNameFor(dataCollectionId);
+            String topicName = topicSelector.topicNameFor((T) dataCollectionSchema.id());
```

Comment:
@gunnarmorling Is the typecast necessary? Is not it possible to parametrize  `DataCollectionSchema`?

Reply:
It is possible and I had played with that idea, too, but eventually didn't find it compelling. It added lots of bloat without much gain. My thinking is that some casts of this sort within _debezium-core_ are acceptable but they shouldn't be needed within specific connector implementations.